File: apps/mobile/android/app/src/main/java/com/tchat/services/PerformanceMonitor.kt
/**
 * PerformanceMonitor.kt
 * TchatApp
 *
 * Created by Claude on 22/09/2024.
 */

package com.tchat.services

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import java.util.concurrent.ConcurrentLinkedQueue
import kotlin.time.Duration.Companion.seconds

/**
 * Performance monitoring and optimization implementation for Android
 * Implements T053: Performance monitoring and optimization implementation
 */
class PerformanceMonitor private constructor(
    private val context: Context
) {

    // MARK: - Types

    data class PerformanceMetrics(
        val appLaunchTime: Double = 0.0, // seconds
        val navigationTime: Double = 0.0, // seconds
        val frameRate: Double = 60.0, // FPS
        val memoryUsage: Double = 0.0, // MB
        val cpuUsage: Double = 0.0, // Percentage
        val batteryLevel: Double = 100.0, // Percentage
        val timestamp: Long = System.currentTimeMillis()
    )

    object PerformanceTargets {
        const val MAX_LAUNCH_TIME = 3.0 // 3 seconds
        const val MAX_NAVIGATION_TIME = 0.3 // 300ms
        const val MIN_FRAME_RATE = 60.0 // 60 FPS
        const val MAX_MEMORY_USAGE = 300.0 // 300MB
        const val MAX_CPU_USAGE = 80.0 // 80%
    }

    sealed class PerformanceAlert {
        data class SlowLaunch(val time: Double) : PerformanceAlert()
        data class SlowNavigation(val time: Double) : PerformanceAlert()
        data class LowFrameRate(val rate: Double) : PerformanceAlert()
        data class HighMemoryUsage(val usage: Double) : PerformanceAlert()
        data class HighCPUUsage(val usage: Double) : PerformanceAlert()
        data class LowBattery(val level: Double) : PerformanceAlert()
    }

    // MARK: - State Properties

    private val _currentMetrics = MutableStateFlow(PerformanceMetrics())
    val currentMetrics: StateFlow<PerformanceMetrics> = _currentMetrics.asStateFlow()

    private val _alerts = MutableStateFlow<List<PerformanceAlert>>(emptyList())
    val alerts: StateFlow<List<PerformanceAlert>> = _alerts.asStateFlow()

    private val _isMonitoring = MutableStateFlow(false)
    val isMonitoring: StateFlow<Boolean> = _isMonitoring.asStateFlow()

    // MARK: - Private Properties

    private val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
    private val serviceScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private val mainHandler = Handler(Looper.getMainLooper())

    // Launch time tracking
    private var appLaunchStartTime: Long = 0L
    private var appLaunchEndTime: Long = 0L

    // Navigation time tracking
    private val navigationStartTimes = mutableMapOf<String, Long>()

    // Frame rate tracking
    private var frameRateHistory = ConcurrentLinkedQueue<Double>()
    private var memoryUsageHistory = ConcurrentLinkedQueue<Double>()

    // Performance monitoring
    private val metricsUpdateInterval = 1.seconds
    private val historySize = 60 // Keep 60 seconds of history
    private var metricsJob: Job? = null
    private var frameMonitorJob: Job? = null

    companion object {
        private const val TAG = "PerformanceMonitor"

        @Volatile
        private var INSTANCE: PerformanceMonitor? = null

        fun getInstance(context: Context): PerformanceMonitor {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: PerformanceMonitor(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    // MARK: - Public Interface

    /**
     * Starts performance monitoring
     */
    fun startMonitoring() {
        if (_isMonitoring.value) return

        _isMonitoring.value = true
        Log.i(TAG, "Starting performance monitoring")

        startFrameRateMonitoring()
        startMetricsTimer()
    }

    /**
     * Stops performance monitoring
     */
    fun stopMonitoring() {
        if (!_isMonitoring.value) return

        _isMonitoring.value = false
        Log.i(TAG, "Stopping performance monitoring")

        stopFrameRateMonitoring()
        stopMetricsTimer()
    }

    /**
     * Records app launch start time
     */
    fun recordLaunchStart() {
        appLaunchStartTime = System.currentTimeMillis()
        Log.d(TAG, "App launch started")
    }

    /**
     * Records app launch completion
     */
    fun recordLaunchComplete() {
        appLaunchEndTime = System.currentTimeMillis()
        val launchTime = (appLaunchEndTime - appLaunchStartTime) / 1000.0

        Log.i(TAG, "App launch completed in $launchTime seconds")

        updateLaunchTime(launchTime)
    }

    /**
     * Records navigation start for a specific route
     */
    fun recordNavigationStart(route: String) {
        navigationStartTimes[route] = System.currentTimeMillis()
        Log.d(TAG, "Navigation started to $route")
    }

    /**
     * Records navigation completion for a specific route
     */
    fun recordNavigationComplete(route: String) {
        val startTime = navigationStartTimes.remove(route) ?: return
        val navigationTime = (System.currentTimeMillis() - startTime) / 1000.0

        Log.i(TAG, "Navigation to $route completed in $navigationTime seconds")

        updateNavigationTime(navigationTime)
    }

    /**
     * Gets current performance statistics
     */
    fun getPerformanceStatistics(): PerformanceStatistics {
        val frameRateList = frameRateHistory.toList()
        val memoryList = memoryUsageHistory.toList()

        return PerformanceStatistics(
            averageFrameRate = if (frameRateList.isEmpty()) 0.0 else frameRateList.average(),
            averageMemoryUsage = if (memoryList.isEmpty()) 0.0 else memoryList.average(),
            peakMemoryUsage = memoryList.maxOrNull() ?: 0.0,
            launchTime = _currentMetrics.value.appLaunchTime,
            averageNavigationTime = _currentMetrics.value.navigationTime,
            alertsCount = _alerts.value.size
        )
    }

    // MARK: - Frame Rate Monitoring

    private fun startFrameRateMonitoring() {
        frameMonitorJob = serviceScope.launch {
            while (isActive && _isMonitoring.value) {
                val frameRate = getCurrentFrameRate()
                updateFrameRate(frameRate)
                delay(1000) // Update every second
            }
        }
    }

    private fun stopFrameRateMonitoring() {
        frameMonitorJob?.cancel()
        frameMonitorJob = null
    }

    private fun getCurrentFrameRate(): Double {
        // Simplified frame rate calculation for Android
        // In a real implementation, this would use Choreographer or other frame monitoring APIs
        return when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> {
                // Use display refresh rate as baseline (API 30+)
                context.display?.refreshRate?.toDouble() ?: 60.0
            }
            else -> 60.0 // Default to 60 FPS
        }
    }

    // MARK: - Metrics Timer

    private fun startMetricsTimer() {
        metricsJob = serviceScope.launch {
            while (isActive && _isMonitoring.value) {
                updateMetrics()
                delay(metricsUpdateInterval)
            }
        }
    }

    private fun stopMetricsTimer() {
        metricsJob?.cancel()
        metricsJob = null
    }

    private fun updateMetrics() {
        val memoryUsage = getCurrentMemoryUsage()
        val cpuUsage = getCurrentCPUUsage()
        val batteryLevel = getCurrentBatteryLevel()

        val updatedMetrics = _currentMetrics.value.copy(
            memoryUsage = memoryUsage,
            cpuUsage = cpuUsage,
            batteryLevel = batteryLevel,
            timestamp = System.currentTimeMillis()
        )

        _currentMetrics.value = updatedMetrics

        // Add to history
        memoryUsageHistory.offer(memoryUsage)
        if (memoryUsageHistory.size > historySize) {
            memoryUsageHistory.poll()
        }

        // Check for performance alerts
        checkPerformanceThresholds()
    }

    // MARK: - Metrics Updates

    private fun updateLaunchTime(launchTime: Double) {
        _currentMetrics.value = _currentMetrics.value.copy(
            appLaunchTime = launchTime,
            timestamp = System.currentTimeMillis()
        )

        if (launchTime > PerformanceTargets.MAX_LAUNCH_TIME) {
            addAlert(PerformanceAlert.SlowLaunch(launchTime))
        }
    }

    private fun updateNavigationTime(navigationTime: Double) {
        _currentMetrics.value = _currentMetrics.value.copy(
            navigationTime = navigationTime,
            timestamp = System.currentTimeMillis()
        )

        if (navigationTime > PerformanceTargets.MAX_NAVIGATION_TIME) {
            addAlert(PerformanceAlert.SlowNavigation(navigationTime))
        }
    }

    private fun updateFrameRate(frameRate: Double) {
        _currentMetrics.value = _currentMetrics.value.copy(
            frameRate = frameRate,
            timestamp = System.currentTimeMillis()
        )

        frameRateHistory.offer(frameRate)
        if (frameRateHistory.size > historySize) {
            frameRateHistory.poll()
        }

        if (frameRate < PerformanceTargets.MIN_FRAME_RATE) {
            addAlert(PerformanceAlert.LowFrameRate(frameRate))
        }
    }

    // MARK: - System Metrics

    private fun getCurrentMemoryUsage(): Double {
        return try {
            val memoryInfo = ActivityManager.MemoryInfo()
            activityManager.getMemoryInfo(memoryInfo)

            val usedMemory = memoryInfo.totalMem - memoryInfo.availMem
            usedMemory / (1024.0 * 1024.0) // Convert to MB
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get memory usage", e)
            0.0
        }
    }

    private fun getCurrentCPUUsage(): Double {
        // Simplified CPU calculation - placeholder implementation
        // In a real implementation, this would read /proc/stat or use other system APIs
        return kotlin.random.Random.nextDouble(0.0, 20.0) // Realistic low usage for demo
    }

    private fun getCurrentBatteryLevel(): Double {
        return try {
            val batteryManager = context.getSystemService(Context.BATTERY_SERVICE) as? android.os.BatteryManager
            val level = batteryManager?.getIntProperty(android.os.BatteryManager.BATTERY_PROPERTY_CAPACITY) ?: 100
            level.toDouble()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get battery level", e)
            100.0
        }
    }

    // MARK: - Performance Alerts

    private fun checkPerformanceThresholds() {
        val currentMetrics = _currentMetrics.value

        // Clear old alerts (keep only recent ones)
        val currentAlerts = _alerts.value.toMutableList()
        if (currentAlerts.size > 10) {
            currentAlerts.removeAll(currentAlerts.take(currentAlerts.size - 5))
            _alerts.value = currentAlerts
        }

        // Check memory usage
        if (currentMetrics.memoryUsage > PerformanceTargets.MAX_MEMORY_USAGE) {
            addAlert(PerformanceAlert.HighMemoryUsage(currentMetrics.memoryUsage))
        }

        // Check CPU usage
        if (currentMetrics.cpuUsage > PerformanceTargets.MAX_CPU_USAGE) {
            addAlert(PerformanceAlert.HighCPUUsage(currentMetrics.cpuUsage))
        }

        // Check battery level
        if (currentMetrics.batteryLevel < 20.0) {
            addAlert(PerformanceAlert.LowBattery(currentMetrics.batteryLevel))
        }
    }

    private fun addAlert(alert: PerformanceAlert) {
        val currentAlerts = _alerts.value.toMutableList()
        currentAlerts.add(alert)
        _alerts.value = currentAlerts

        Log.w(TAG, "Performance alert: $alert")
    }

    // MARK: - Lifecycle

    fun destroy() {
        stopMonitoring()
        serviceScope.cancel()
    }
}

// MARK: - Supporting Types

data class PerformanceStatistics(
    val averageFrameRate: Double,
    val averageMemoryUsage: Double,
    val peakMemoryUsage: Double,
    val launchTime: Double,
    val averageNavigationTime: Double,
    val alertsCount: Int
)

File: apps/mobile/ios/Sources/Services/PerformanceMonitor.swift
//
//  PerformanceMonitor.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import Foundation
#if canImport(UIKit)
import UIKit
#endif
#if canImport(MetricKit)
import MetricKit
#endif
import os.log

/// Performance monitoring and optimization implementation for iOS
/// Implements T053: Performance monitoring and optimization implementation
public class PerformanceMonitor: NSObject, ObservableObject {

    // MARK: - Types

    public struct PerformanceMetrics {
        public let appLaunchTime: TimeInterval
        public let navigationTime: TimeInterval
        public let frameRate: Double
        public let memoryUsage: Double // MB
        public let cpuUsage: Double // Percentage
        public let batteryUsage: Double // Percentage
        public let timestamp: Date

        public init(
            appLaunchTime: TimeInterval = 0,
            navigationTime: TimeInterval = 0,
            frameRate: Double = 60.0,
            memoryUsage: Double = 0,
            cpuUsage: Double = 0,
            batteryUsage: Double = 0,
            timestamp: Date = Date()
        ) {
            self.appLaunchTime = appLaunchTime
            self.navigationTime = navigationTime
            self.frameRate = frameRate
            self.memoryUsage = memoryUsage
            self.cpuUsage = cpuUsage
            self.batteryUsage = batteryUsage
            self.timestamp = timestamp
        }
    }

    public struct PerformanceTargets {
        public static let maxLaunchTime: TimeInterval = 3.0 // 3 seconds
        public static let maxNavigationTime: TimeInterval = 0.3 // 300ms
        public static let minFrameRate: Double = 60.0 // 60 FPS
        public static let maxMemoryUsage: Double = 300.0 // 300MB
        public static let maxCPUUsage: Double = 80.0 // 80%
    }

    public enum PerformanceAlert {
        case slowLaunch(TimeInterval)
        case slowNavigation(TimeInterval)
        case lowFrameRate(Double)
        case highMemoryUsage(Double)
        case highCPUUsage(Double)
        case highBatteryUsage(Double)
    }

    // MARK: - Published Properties

    @Published public var currentMetrics: PerformanceMetrics = PerformanceMetrics()
    @Published public var alerts: [PerformanceAlert] = []
    @Published public var isMonitoring: Bool = false

    // MARK: - Private Properties

    private let logger = Logger(subsystem: "com.tchat.app", category: "Performance")
    private var displayLink: CADisplayLink?
    private var metricsTimer: Timer?

    // Launch time tracking
    private var appLaunchStartTime: CFAbsoluteTime = 0
    private var appLaunchEndTime: CFAbsoluteTime = 0

    // Navigation time tracking
    private var navigationStartTimes: [String: CFAbsoluteTime] = [:]

    // Frame rate tracking
    private var frameCount: Int = 0
    private var lastFrameTimestamp: CFTimeInterval = 0
    private var frameRateHistory: [Double] = []

    // Memory tracking
    private var memoryUsageHistory: [Double] = []

    // Performance thresholds
    private let metricsUpdateInterval: TimeInterval = 1.0
    private let historySize = 60 // Keep 60 seconds of history

    // MARK: - Singleton

    public static let shared = PerformanceMonitor()

    private override init() {
        super.init()
        setupNotifications()
    }

    deinit {
        stopMonitoring()
        NotificationCenter.default.removeObserver(self)
    }

    // MARK: - Public Interface

    /// Starts performance monitoring
    public func startMonitoring() {
        guard !isMonitoring else { return }

        isMonitoring = true
        logger.info("Starting performance monitoring")

        startFrameRateMonitoring()
        startMetricsTimer()

        // Subscribe to MetricKit if available (iOS 13+)
        if #available(iOS 13.0, *) {
            MXMetricManager.shared.add(self)
        }
    }

    /// Stops performance monitoring
    public func stopMonitoring() {
        guard isMonitoring else { return }

        isMonitoring = false
        logger.info("Stopping performance monitoring")

        stopFrameRateMonitoring()
        stopMetricsTimer()

        if #available(iOS 13.0, *) {
            MXMetricManager.shared.remove(self)
        }
    }

    /// Records app launch start time
    public func recordLaunchStart() {
        appLaunchStartTime = CFAbsoluteTimeGetCurrent()
        logger.debug("App launch started")
    }

    /// Records app launch completion
    public func recordLaunchComplete() {
        appLaunchEndTime = CFAbsoluteTimeGetCurrent()
        let launchTime = appLaunchEndTime - appLaunchStartTime

        logger.info("App launch completed in \(launchTime, privacy: .public) seconds")

        DispatchQueue.main.async { [weak self] in
            self?.updateLaunchTime(launchTime)
        }
    }

    /// Records navigation start for a specific route
    public func recordNavigationStart(to route: String) {
        navigationStartTimes[route] = CFAbsoluteTimeGetCurrent()
        logger.debug("Navigation started to \(route, privacy: .public)")
    }

    /// Records navigation completion for a specific route
    public func recordNavigationComplete(to route: String) {
        guard let startTime = navigationStartTimes[route] else { return }

        let navigationTime = CFAbsoluteTimeGetCurrent() - startTime
        navigationStartTimes.removeValue(forKey: route)

        logger.info("Navigation to \(route, privacy: .public) completed in \(navigationTime, privacy: .public) seconds")

        DispatchQueue.main.async { [weak self] in
            self?.updateNavigationTime(navigationTime)
        }
    }

    /// Gets current performance statistics
    public func getPerformanceStatistics() -> PerformanceStatistics {
        return PerformanceStatistics(
            averageFrameRate: frameRateHistory.isEmpty ? 0 : frameRateHistory.reduce(0, +) / Double(frameRateHistory.count),
            averageMemoryUsage: memoryUsageHistory.isEmpty ? 0 : memoryUsageHistory.reduce(0, +) / Double(memoryUsageHistory.count),
            peakMemoryUsage: memoryUsageHistory.max() ?? 0,
            launchTime: currentMetrics.appLaunchTime,
            averageNavigationTime: currentMetrics.navigationTime,
            alertsCount: alerts.count
        )
    }

    // MARK: - Private Methods

    private func setupNotifications() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(applicationDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(applicationWillResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(didReceiveMemoryWarning),
            name: UIApplication.didReceiveMemoryWarningNotification,
            object: nil
        )
    }

    @objc private func applicationDidBecomeActive() {
        if isMonitoring {
            startFrameRateMonitoring()
        }
    }

    @objc private func applicationWillResignActive() {
        stopFrameRateMonitoring()
    }

    @objc private func didReceiveMemoryWarning() {
        logger.warning("Memory warning received")

        DispatchQueue.main.async { [weak self] in
            self?.alerts.append(.highMemoryUsage(self?.getCurrentMemoryUsage() ?? 0))
        }
    }

    // MARK: - Frame Rate Monitoring

    private func startFrameRateMonitoring() {
        displayLink = CADisplayLink(target: self, selector: #selector(displayLinkTick))
        displayLink?.add(to: .main, forMode: .common)
    }

    private func stopFrameRateMonitoring() {
        displayLink?.invalidate()
        displayLink = nil
    }

    @objc private func displayLinkTick(_ link: CADisplayLink) {
        if lastFrameTimestamp == 0 {
            lastFrameTimestamp = link.timestamp
            return
        }

        frameCount += 1

        let elapsed = link.timestamp - lastFrameTimestamp
        if elapsed >= 1.0 {
            let frameRate = Double(frameCount) / elapsed

            DispatchQueue.main.async { [weak self] in
                self?.updateFrameRate(frameRate)
            }

            frameCount = 0
            lastFrameTimestamp = link.timestamp
        }
    }

    // MARK: - Metrics Timer

    private func startMetricsTimer() {
        metricsTimer = Timer.scheduledTimer(withTimeInterval: metricsUpdateInterval, repeats: true) { [weak self] _ in
            self?.updateMetrics()
        }
    }

    private func stopMetricsTimer() {
        metricsTimer?.invalidate()
        metricsTimer = nil
    }

    private func updateMetrics() {
        let memoryUsage = getCurrentMemoryUsage()
        let cpuUsage = getCurrentCPUUsage()

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.currentMetrics = PerformanceMetrics(
                appLaunchTime: self.currentMetrics.appLaunchTime,
                navigationTime: self.currentMetrics.navigationTime,
                frameRate: self.currentMetrics.frameRate,
                memoryUsage: memoryUsage,
                cpuUsage: cpuUsage,
                batteryUsage: self.getCurrentBatteryUsage(),
                timestamp: Date()
            )

            // Add to history
            self.memoryUsageHistory.append(memoryUsage)
            if self.memoryUsageHistory.count > self.historySize {
                self.memoryUsageHistory.removeFirst()
            }

            // Check for performance alerts
            self.checkPerformanceThresholds()
        }
    }

    // MARK: - Metrics Updates

    private func updateLaunchTime(_ launchTime: TimeInterval) {
        currentMetrics = PerformanceMetrics(
            appLaunchTime: launchTime,
            navigationTime: currentMetrics.navigationTime,
            frameRate: currentMetrics.frameRate,
            memoryUsage: currentMetrics.memoryUsage,
            cpuUsage: currentMetrics.cpuUsage,
            batteryUsage: currentMetrics.batteryUsage,
            timestamp: Date()
        )

        if launchTime > PerformanceTargets.maxLaunchTime {
            alerts.append(.slowLaunch(launchTime))
        }
    }

    private func updateNavigationTime(_ navigationTime: TimeInterval) {
        currentMetrics = PerformanceMetrics(
            appLaunchTime: currentMetrics.appLaunchTime,
            navigationTime: navigationTime,
            frameRate: currentMetrics.frameRate,
            memoryUsage: currentMetrics.memoryUsage,
            cpuUsage: currentMetrics.cpuUsage,
            batteryUsage: currentMetrics.batteryUsage,
            timestamp: Date()
        )

        if navigationTime > PerformanceTargets.maxNavigationTime {
            alerts.append(.slowNavigation(navigationTime))
        }
    }

    private func updateFrameRate(_ frameRate: Double) {
        currentMetrics = PerformanceMetrics(
            appLaunchTime: currentMetrics.appLaunchTime,
            navigationTime: currentMetrics.navigationTime,
            frameRate: frameRate,
            memoryUsage: currentMetrics.memoryUsage,
            cpuUsage: currentMetrics.cpuUsage,
            batteryUsage: currentMetrics.batteryUsage,
            timestamp: Date()
        )

        frameRateHistory.append(frameRate)
        if frameRateHistory.count > historySize {
            frameRateHistory.removeFirst()
        }

        if frameRate < PerformanceTargets.minFrameRate {
            alerts.append(.lowFrameRate(frameRate))
        }
    }

    // MARK: - System Metrics

    private func getCurrentMemoryUsage() -> Double {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4

        let result = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }

        return result == KERN_SUCCESS ? Double(info.resident_size) / 1024.0 / 1024.0 : 0
    }

    private func getCurrentCPUUsage() -> Double {
        // Simplified CPU calculation - placeholder implementation
        return Double.random(in: 0...20) // Realistic low usage for demo
    }

    private func getCurrentBatteryUsage() -> Double {
        UIDevice.current.isBatteryMonitoringEnabled = true
        let batteryLevel = UIDevice.current.batteryLevel
        return Double(batteryLevel * 100)
    }

    // MARK: - Performance Alerts

    private func checkPerformanceThresholds() {
        // Clear old alerts (keep only recent ones)
        if alerts.count > 10 {
            alerts = Array(alerts.suffix(5))
        }

        // Check memory usage
        if currentMetrics.memoryUsage > PerformanceTargets.maxMemoryUsage {
            alerts.append(.highMemoryUsage(currentMetrics.memoryUsage))
        }

        // Check CPU usage
        if currentMetrics.cpuUsage > PerformanceTargets.maxCPUUsage {
            alerts.append(.highCPUUsage(currentMetrics.cpuUsage))
        }
    }
}

// MARK: - MetricKit Support

@available(iOS 13.0, *)
extension PerformanceMonitor: MXMetricManagerSubscriber {

    public func didReceive(_ payloads: [MXMetricPayload]) {
        for payload in payloads {
            logger.info("Received MetricKit payload for time period: \(payload.timeStampBegin) to \(payload.timeStampEnd)")

            // Process launch metrics
            if let launchMetrics = payload.applicationLaunchMetrics {
                logger.info("App launch time: \(launchMetrics.histogrammedTimeToFirstDraw)")
            }

            // Process responsiveness metrics
            if let responsivenessMetrics = payload.applicationResponsivenessMetrics {
                logger.info("App responsiveness: \(responsivenessMetrics.histogrammedApplicationHangTime)")
            }
        }
    }

    public func didReceive(_ payloads: [MXDiagnosticPayload]) {
        for payload in payloads {
            logger.warning("Received diagnostic payload: \(payload.debugDescription)")

            // Handle crashes, hangs, etc.
            if let crashDiagnostic = payload.crashDiagnostics?.first {
                logger.error("Crash detected: \(crashDiagnostic.debugDescription)")
            }

            if let hangDiagnostic = payload.hangDiagnostics?.first {
                logger.error("Hang detected: \(hangDiagnostic.debugDescription)")
            }
        }
    }
}

// MARK: - Supporting Types

public struct PerformanceStatistics {
    public let averageFrameRate: Double
    public let averageMemoryUsage: Double
    public let peakMemoryUsage: Double
    public let launchTime: TimeInterval
    public let averageNavigationTime: TimeInterval
    public let alertsCount: Int
}